pkg/gomr/utils.go

```diff
@@ -69,6 +70,42 @@ func Match(line, regex string) bool {
 	return matchRegex.MatchString(line)
 }
 
+// CanonicalizeIrcNick converts a string to its downcased version according to
+// the IRC protocol, with invalid characters removed.
+func CanonicalizeIrcNick(nick string) string {
+	// Strip invalid characters and convert to lowercase.
+	nick = strings.Map(func(r rune) rune {
+		switch {
+		case r == '[':
+			// '[' is uppercase '{'.
+			return '{'
+		case r == ']':
+			// ']' is uppercase '}'.
+			return '}'
+		case r == '\\':
+			// '\' is uppercase '|'.
+			return '|'
+		case r == '^' || r == '_' || r == '`' || r == '{' || r == '\\' || r == '}':
+			// These characters are fine.
+			return r
+		case r >= 'A' && r <= 'Z':
+			// Make uppercase letters lowercase.
+			return r + 32
+		case r >= 'a' && r <= 'z' || r >= '0' && r <= '9':
+			// Lowercase letters and numerals are fine.
+			return r
+		}
+		return -1
+	}, nick)
+
+	// Strip all invalid leading characters that are otherwise valid.
+	nick = strings.TrimLeftFunc(nick, func(r rune) bool {
+		return (r >= '0' && r <= '9') || r == '-'
```

Comment:
Numbers are invalid leading characters? What if I'm a big fan of 3scale++?

Reply:
https://tools.ietf.org/html/rfc2812#page-7 says it's invalid.  That doesn't mean no server allows it though.

Reply:
Ohh, I see; we don't want to restrict this to IRC nicks.  I'll loosen the restrictions, maybe just drop control characters and convert to lowercase (per how IRC defines lowercase).

Reply:
Fixed.